in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+    const outerHash = computeOuterAuthWitHash(
+      this.dappEntrypointAddress,
+      new Fr(this.chainId),
+      new Fr(this.version),
+      innerHash,
+    );
+    const authWitness = await this.userAuthWitnessProvider.createAuthWit(outerHash);
+
+    builder
+      .setOrigin(this.dappEntrypointAddress)
+      .setTxContext({
+        chainId: this.chainId,
+        version: this.version,
+        gasSettings: userRequest.gasSettings,
+      })
+      .setFunctionSelector(functionSelector)
+      .setFirstCallArgsHash(entrypointPackedArgs.hash)
+      .setArgsOfCalls([...payload.packedArguments, entrypointPackedArgs])
+      .addAuthWitness(authWitness);
+
+  }
+
+
+  createAuthWit(messageHash: Fr): Promise<AuthWitness> {
+    return this.authWitnessProvider.createAuthWit(messageHash);
+  }
+
+  getCompleteAddress(): CompleteAddress {
+    return this.completeAddress;
+  }
+
+  getAddress(): AztecAddress {
+    return this.completeAddress.address;
+  }
+
+  getChainId(): Fr {
+    return this.chainId;
+  }
+
+  getVersion(): Fr {
+    return this.version;
+  }
+
+  private getEntrypointAbi() {
+    return {
+      name: 'entrypoint',
+      // ... same as before
+    }
+  }
+
+}
+```
+
+#### Create the wallet as normal
+
+```ts
+export async function getDappWallet(
+  pxe: PXE,
+  accountAddress: AztecAddress,
+  dappAddress: AztecAddress,
+  userAuthWitnessProvider: AuthWitnessProvider,
+): Promise<AccountWallet> {
+  const completeAddress = await pxe.getRegisteredAccount(accountAddress);
+  if (!completeAddress) {
+    throw new Error(`Account ${address} not found`);
+  }
+  const nodeInfo = await pxe.getNodeInfo();
+  const entrypoint = new DefaultDappInterface(completeAddress, userAuthWitnessProvider, dappAddress);
+  return new AccountWallet(pxe, entrypoint);
+}
+
+
+const schnorr = new SchnorrAccountContract(signingPrivateKey);
+const authWitProvider = schnorr.getAuthWitnessProvider();
+const aliceDappWrappedWallet = await getDappWallet(pxe, aliceAddress, dappAddress, userAuthWitnessProvider);
+
+const { request: deployAliceAccountRequest } = await aliceDappWrappedWallet.simulate({
+  calls: [{
+    contractInstance: bananaCoinInstance,
+    functionName: 'transfer',
+    args: { 
+      from: aliceAddress,
+      to: bobAddress,
+      value: privateBalance,
+      nonce: 0n
+    },
+  }],
+  paymentMethod: new NoFeePaymentMethod(),
+})
+
+```
+
+
+### Gas Estimation
+
+```ts
+
+
+export interface GasEstimator {
+  proposeGasSettings(): GasSettings;
+  isConverged(): boolean;
+  update(simulationOutput: SimulationOutput): void;
+}
+
+interface SearchRatios {
+  daToL2: number;
+  daTearDown: number;
+  l2TearDown: number;
+}
+
+// Finds a balance that is enough to cover the gas costs of the simulation.
+// Marks as converged if the simulation did not run out of gas,
+// or if it is not possible to increase the balance further.
+export class BinarySearchGasEstimator implements GasEstimator {
```

Comment:
Not super into how you are doing much gas estimation, so I will mostly ignore this for now.